### PR TITLE
Deprecate zope.component in favor of an direct calls to functions from `certbot.display.util` module

### PR DESCRIPTION
**Pull request by adferrand**
This is the first PR of a serie with the goal of removing totally zope dependencies from Certbot. As discussed previously with @bmw, the steps will be the following:
1) Removing mechanics around `zope.component`
2) Creating ABC classes that will replace Zope interfaces from `certbot.intefaces`
3) Rewrite plugin discovery logic to use the ABC classes instead of the Zope interfaces
4) Removing Zope interfaces, `zope.*` dependencies and deprecation paths added during the previous steps

This PR is solving the first step.

`zope.component` is used in Certbot to provide singleton logic: global objects implementing `IConfig`, `IReporter` and `IDisplay` need to be available everywhere in Certbot code, and `provideUtility`/`getUtility` from `zope.component` are used for that purpose.

In this PR a new module `certbot.services` is created, and reimplements this singleton logic through the getters/setters module functions for config, reporter and display. All Certbot code is adapted to use this new logic.

The `certbot.services` is part of the public API.

A deprecation path is provided for `provideUtility` and `getUtility` through the init module of `certbot`: when these functions are called for the Certbot Zope interfaces, a warning is emitted and the call is delegated to the `certbot.services` module that will return the actual singletons. This way external plugins maintainers have the occasion to be warned about deprecated usages of `zope.component` for Certbot and the necessity to move to the new logic.

Beside obvious changes throughout the Certbot code, an extra care was needed for `certbot.tests.util.patch_get_utility`. Indeed despite this decorator claiming to patch only the display utility, it was actually patching any output of `getUtility` function, including reporter and config, even though the mock objects specs were taken from the display utility interface... To make things clear, I make `patch_get_utility` explicitly mock display only, and change its name appropriately. In test modules either reporter or display are mocked as needed.

### CODE REVIEW

**Comment by adferrand** [comment]
For whoever is interested to see what I plan for the second step of this work, you can see  the ongoing work on the following PR, which builds the new interfaces on top of the step 1: https://github.com/adferrand/certbot/pull/73

**Comment by bmw** [comment]
Thanks a lot for picking this up! I'm glad to finally be removing our dependency on `zope`.

I had an idea for an alternate design here that I wanted hear your thoughts on. It involves some additional refactoring, but I'd be happy to try and help with this if you like the idea to avoid pushing all the additional work on you. Essentially the idea is to deprecate the config and reporter singletons completely by requiring the config/reporter to be passed around and to implement `IDisplay` as a Python module with functions and private global state.

I think we can completely deprecate the config and reporter singletons because there are only 3 uses of `services.get_config` in this PR outside of tests and [after the UX PRs are merged, `IReporter` is no longer used](https://opensource.eff.org/eff-open-source/pl/t3ma6ntjt7dszxzzxj5weup6ay). 2 of the `services.get_config` uses are in functions that are part of Certbot's public API which complicates this, but I think that's a solvable problem and I like reducing the amount of global state.

As for `IDisplay`, I was motivated by posts like https://python-patterns.guide/gang-of-four/singleton/ and https://stackoverflow.com/a/31887/6534779 which suggests using a Python module with functions and private global state to implement singletons. I think this approach can result in less/simpler code, provides a cleaner API (`display.method` as opposed to `get_display().method`), and removes the need for us to define a separate display interface in the long term which I think would resolve https://github.com/certbot/certbot/issues/8434 because I think the documentation for the interface of the module is sufficient.

There are some minor details here to be worked out such as additional methods for setting global attributes like where output should be written, but I'm not foreseeing any major hurdles here. I also think this approach can be used to satisfy `zope` temporarily which I tested successfully with [this patch](https://gist.github.com/5d812574bfaf6fca3bd3bd813b70161e).

What do you think? Are there any aspects of this idea that you don't like? If you're on board, would you like any help implementing it?

**Comment by adferrand** [comment]
I implemented a service module handling the three modules for the following reasons:
* it is the closest implementation to the current situation with `zope.component`
* thanks to that the deprecation path can use bijective replacements with dedicated warning to prepare the community
* it does not modify the public API

If you look closely at the services module, you will see that it follows in fact the approach you describe about functions on a module and an internal state, but one level up since the module holds the three services instead of having one module per service.

Indeed I agree that if one service will be removed anyway, and one is effectively quite simple to pass as standard function arguments, it is really interesting in term of code quality to move the module singleton pattern to the remaining display service.

However I fear that at that point it will be difficult to create a clear deprecation system,  given that in this case zope.component.getUtility cannot work anymore for `IConfig`. It will also add more divergence on the public API on top of my next PR for `zope.interface`, where the public API is totally inevitable. 

So I would say that it depends on the ambition we have about public API changes for the CLI refactoring.

If it is significant, I am totally OK with your proposition since the deprecation system and the public API changes for this feature will not really matter in this case. 

If it is not, I prefer to keep with my design that minimize the impact, and implement your improvements when we will remove the deprecated features.

**Comment by bmw** [comment]
> I fear that at that point it will be difficult to create a clear deprecation system,  given that in this case zope.component.getUtility cannot work anymore for `IConfig`.

What if we added the deprecation message to references of the interfaces like we did with [tls-sni](https://github.com/certbot/certbot/pull/6859)? Any time a user touches a deprecated interface, we raise a warning which we probably should do anyway to deprecate the `zope` interfaces and it allows us to remove the monkey patching of `zope.component`.

I think another option would be effectively wrapping `zope.component.getUtility` in the monkey patch by raising warnings on calls for deprecated interfaces before delegating the call to the real `zope.component.getUtility` function, but I personally like the first option better if we can make it work.

> So I would say that it depends on the ambition we have about public API changes for the CLI refactoring.
>
> If it is significant, I am totally OK with your proposition since the deprecation system and the public API changes for this feature will not really matter in this case.
>
> If it is not, I prefer to keep with my design that minimize the impact, and implement your improvements when we will remove the deprecated features.

I'm not sure I understand what you're saying here. Are you saying that if the UX changes Alex and Erica have been working on involve significant API changes, you're fine making more here, but if they do not, you'd like to try and minimize the API changes made with this PR? What deprecated features are you referring to when you say "I prefer to keep with my design that minimize the impact, and implement your improvements when we will remove the deprecated features."?

My personal problem is I think that deprecating `zope` necessitates API changes, I personally like the design in my previous post better in the long term, and I'd rather not transition ourselves and our users to a temporary API if that's how we'd eventually like things to work.

Let me know if you'd prefer to have the rest of this conversation over Mattermost or a video call.

**Comment by adferrand** [comment]
As discussed above and on Mattermost, this PR implements now a slightly improved approach:
* remove completely calls to `IReporter`
* `certbot.display.util` now holds a global singleton of `IDisplay`
* `certbot.display.util` exposes to the outside world functions that delegate to the underlying `IDisplay` singleton

As a result, certbot codebase uses the display utilities only through module functions. I have kept calls to `zope.component.provideUtility` for compatibility purpose. The overall deprecation of these mechanics will be done with the deprecation of `zope.interface` in Certbot.

This PR is ready for a new review.

**Comment by adferrand** [comment]
NB: The failure of sphinx builds will be fixed with #8863

File: certbot/certbot/display/util.py
Comment (bmw) [inline certbot/certbot/display/util.py:204]: I wonder if we could and should make this function private? Why should the user of this module ever access the display object instead of using the module level functions? Looking through the code, I found only two use cases:

1. Unit tests where I think we could mock all of `certbot.display.util`. (Since we tell people to run our tests with `pytest`, I'm also interested in doing something like setting up a `context.py` file which creates a simple display object when running tests to remove the need to use mocks here in some cases, however, I think this is out of the scope of this PR.)
2. The code [here](https://github.com/certbot/certbot/blob/c372dd8aeef509c193eb1d4615c39b5860b81293/certbot/certbot/_internal/cli/helpful.py#L68-L74) where `self.notify` is used [here](https://github.com/certbot/certbot/blob/c372dd8aeef509c193eb1d4615c39b5860b81293/certbot/certbot/_internal/cli/helpful.py#L154-L157). I think the only time `self.notify` is used, `zope.component.getUtility` fails because we're parsing arguments before setting up the display and noticing that the user just requested `--help` information which we print and exit. I think we could probably either replace the usage of `self.notify` with `print` or set up a temporary display that can be used until arguments are parsed.

What do you think? Do you think these changes are worth it? If so, I could help make them if you like.

**Review by bmw — COMMENTED** [review]
This isn't a full review, but it's a start and I figured I'd post what I have in case you have a chance to take a look.

File: certbot/certbot/display/util.py
Comment (adferrand) [reply certbot/certbot/display/util.py:204]: Yes I hesitated to make it private. Ideally, we should always access the underlying display object through the module functions. Case 1 is not a big deal, 2 was a little more tricky. My intention was indeed to use a dedicate display object for `self.notify`, but could not find an approach that ensures I did not missed some corner case:
* either we set the display object by `set_display`, but in this case invoking the `HelpArgumentParser` has the side effect to define globally a display without notice
* either we use a local display object but we do not benefit from an existing display that may potentially be set before we create the `HelpArgumentParser` instance.

Now if you say that one or all these situations are not a problem, I can update accordingly the code so that we do not need to use `get_display()`.

File: certbot/certbot/display/util.py
Comment (adferrand) [reply certbot/certbot/display/util.py:51]: For sure. This code was more an adaptation of the previous approach. We can certainly use directly a variable now.

File: certbot/certbot/display/util.py
Comment (bmw) [reply certbot/certbot/display/util.py:204]: What do you think about something like:
```
diff --git a/certbot/certbot/_internal/main.py b/certbot/certbot/_internal/main.py
index ed9476c20..c02c2e29f 100644
--- a/certbot/certbot/_internal/main.py
+++ b/certbot/certbot/_internal/main.py
@@ -1511,6 +1511,11 @@ def main(cli_args=None):
     # Some releases of Windows require escape sequences to be enable explicitly
     misc.prepare_virtual_console()
 
+    # HelpfulArgumentParser in prepare_and_parse_args uses display_util so we
+    # create a temporary display that it can use.
+    pre_are_parse_display = display_util.NoninteractiveDisplay(sys.stdout)
+    zope.component.provideUtility(pre_are_parse_display)
+
     # note: arg parser internally handles --help (and exits afterwards)
     args = cli.prepare_and_parse_args(plugins, cli_args)
     config = configuration.NamespaceConfig(args)
```
I'd personally kind of prefer to refactor `HelpfulArgumentParser` so that it doesn't use `display_util`, but I think a change like this is simpler and makes more sense for this PR. This stuff is all internal so we can refactor it later if we really want.

File: certbot/certbot/display/util.py
Comment (adferrand) [reply certbot/certbot/display/util.py:51]: Erratum, I remember another reason. If we use directly a variable, and not the member of an existing object, Python seems to interpret things like `_SERVICE = service` as the local declaration of new variable inside the body of a function (typically here the setter), thus shadowing the module variable `_SERVICE` and so leaving it untouched.

Comment (bmw) [reply certbot/certbot/display/util.py:51]: Ah right. I think we could use the `global` keyword to fix that, but using this class avoids the problem of someone forgetting and accidentally creating a local variable.

I think it'd be nice to add a code comment explaining this, but continuing to use the class is fine with me.

File: certbot/certbot/display/util.py
Comment (adferrand) [reply certbot/certbot/display/util.py:206]: For a few lines of code more, what about creating this display object inside `prepare_and_parse_args`, and pass it as a constructor argument to `HelpfulArgumentParser`?

Comment (bmw) [reply certbot/certbot/display/util.py:206]: That works for me!

File: certbot/certbot/display/util.py
Comment (adferrand) [reply certbot/certbot/display/util.py:53]: Done.

File: certbot/certbot/display/util.py
Comment (adferrand) [reply certbot/certbot/display/util.py:206]: New update. After looking at the code again, I convinced myself that `self.notify` was always set to a dedicated display object and not a preexisting one. Given that `HelpfulArgumentParser` is internal, I took the simplest approach of removing the branch and always create this dedicated display object during the initialization of an instance of `HelpfulArgumentParser`.

Comment (bmw) [reply certbot/certbot/display/util.py:206]: That also works for me! At least when `self.notify` is actually used, I agree it's always a dedicated display object.

With that out of the way, I'd personally like to make `get_display` private to remove it from our API. I think in the medium/long term, we should fix up any tests mocking `get_display` and have them mock `certbot.display.util` instead, but I personally think mocking out a private `_get_display` method temporarily may be fine in some cases for the purposes of getting this PR merged.

Comment (adferrand) [reply certbot/certbot/display/util.py]: This is done.

File: certbot/certbot/tests/util.py
Comment (bmw) [inline certbot/certbot/tests/util.py:202]: I think these functions should continue to patch `zope.component.getUtility` and leave `_get_display` untouched.

Wherever we choose to mock here, I think there's the potential for things to break, but I think the approach of continuing to patch `zope.component.getUtility` will only break if the user of the function was assuming our code would call `getUtility` internally which isn't an assumption I think they should make. If their own code was calling `getUtility`, leaving these functions as is (with the warning you added) will continue to work until they've updated their code to use the new interface.

File: certbot/certbot/tests/util.py
Comment (bmw) [inline certbot/certbot/tests/util.py:183]: I think I was somewhat misleading in my comment at https://github.com/certbot/certbot/pull/8835#discussion_r644970405 so I'm sorry for that.

To be clearer, I'm totally fine with us messing with internals more than we should temporarily, but I think we shouldn't add any code that does that to our public API. Since I don't think we should be messing with internals unless absolutely necessary long term, I'd like to avoid that becoming part of our public API making it harder for us to change.

I'm open to other options, but I personally think the best option for dealing with this is to just bite the bullet and fix things up to patch `display.util` now. I think the main trouble with that will be that the mocks are used slightly differently (e.g. `patched_get_utility().notification` vs `patched_display_util.notification`). If this is done, I'd also like to suggest naming these functions something like `patch_display_util*`.

Alternatively, I think there's a few ways we could temporarily patch `_get_display` in our tests without making code that does it part of our public API, but I don't think we should use internals like that between our packages and my suspicion is that after making all the changes necessary to make things like pylint happy, it'll be almost as much work as just using `display.util` now.

I know this comment is suggesting a fair bit of work, but I hope you understand my concerns here. I'm certainly open to other ideas if you have them.

File: certbot/certbot/display/util.py
Comment (bmw) [inline certbot/certbot/display/util.py:52]:
```suggestion
# This class holds the global state of the display service to avoid the use of
# the "global" keyword when trying to modify it which can be error prone. See
# https://github.com/certbot/certbot/pull/8835/files#r644313793 for more info.
```

**Review by bmw — CHANGES_REQUESTED** [review]
I just reviewed the public API changes here, however, I think reviewing all other changes should be pretty straightforward once the public API is settled.

File: certbot/certbot/display/obj.py
Comment (adferrand) [reply certbot/certbot/display/obj.py:1]: So it means that ultimately (once the imports done for retrocompatiblity in `display.util` are removed), `FileDisplay` and `NonInteractiveDisplay` classes are removed from the public API?

Comment (bmw) [reply certbot/certbot/display/obj.py:1]: Yeah that's what I was thinking. With the exception of `HelpfulArgumentParser` which is also internal and can continue to use the classes, I feel like these classes are now implementation details of `display.util` and it doesn't feel necessary to me to have them in the public API. What do you think?

Comment (adferrand) [reply certbot/certbot/display/obj.py:1]: Well, in this case it seems strange that setting the concrete display with `set_display` is part of the public API, but we do not provide any public concrete implementation of the interface accepted by this setter.

If we go that way, I think we should either provide factories to generate them, like `generate_display(noninteractive=False)`, returning the interface while making the actual instantiation internally. Or we could also make the setter internal and configure the module only through public method, making the display effectively internal (and the public interface useless by the way).

File: certbot/certbot/display/util.py
Comment (adferrand) [reply certbot/certbot/display/util.py:26]: Indeed!

File: certbot/certbot/tests/util.py
Comment (adferrand) [reply certbot/certbot/tests/util.py:183]: No this is on me, I forgot that this module is public.

So I am fine with patching the public part of the module, but I am not sure about the best way to do that. If we want to be able to call something like `mock_util.notification`, we need to mock the entire `display.util`module. It does not seem to be a very common approach with mocks.

Or we patch the public methods themselves, and work with their mock directly. It seems better, because the mock is more accurate. In this case, do we really want to clutter the public API of the `test` module with several `patch_get_utility*`functions? It seems easier to directly call `mock.match("certbot.display.util.method")` in the relevant tests.

File: certbot/certbot/display/obj.py
Comment (bmw) [reply certbot/certbot/display/obj.py:1]: Thanks for pointing out that `set_display` is public. I missed that and I agree that's awkward with private display implementations.

What if we only put functions somewhere in `certbot._internal.display` to get and set the display? The message in the error raised by `get_display` and therefore the `display.util` functions if a display isn't set is just something like:

> This function was called too early in Certbot's execution as the display utility hasn't been configured yet.

I doubt it really matters much, but we could also further specify when the module will be ready by saying something like it's done in `certbot.main.main` after argument parsing.

While I think this design of a utility that gets set up automagically by Certbot internals sometime during startup is a little awkward, I think it buys us the most flexibility. It would mean that users couldn't set or define their own display implementations, but I've never seen anyone do that. https://github.com/certbot/certbot/issues/8434 also exists because Erica and Alex wanted to expand the API of IDisplay but didn't do so because it's technically a breaking change if anyone else was implementing that interface. This would also mean that we wouldn't add a (public) replacement for `IDisplay` as part of ripping out `zope.interface`.

Thoughts?

Comment (bmw) [reply certbot/certbot/display/obj.py:1]: To be clear, I think I've seen 3rd party plugins use the `IDisplay` interface which they could still do with this plan through `display.util`. They just couldn't set or implement their own.

Comment (adferrand) [reply certbot/certbot/display/obj.py:1]: Ok, the plan sounds good to me, if plugins actually do not set themselve the display service.

File: certbot/certbot/tests/util.py
Comment (bmw) [reply certbot/certbot/tests/util.py:202]: Really what I'd like to do here is patch both functions, but I don't see a good way to do that.

File: certbot/certbot/tests/util.py
Comment (bmw) [reply certbot/certbot/tests/util.py:183]: I don't think patching things at the module level is too horrible, but I agree it's a somewhat strange things to do. I also agree it'd probably make the most sense to not add more to our public test API and just patch things as needed in the relevant tests, but that seems like a ton of work and will result in a large PR that does a lot more than just find and replaces a few functions.

I'm honestly struggling a bit to come up with a plan I like here so feel free to propose an alternative, but what about this:

1. In this PR for now, the `get_display` and `set_display` functions continue to use `zope.component`. This is just a temporary measure to simplify the transition so `patch_get_utility*` continues to work on the new functions.
2. In subsequent PR(s), we move our tests away from using `patch_get_utility*`. I think just patching things in the relevant tests makes a lot of sense, however, I also think it may be worth us considering doing something with pytest like I suggested in my parenthetical [here](https://github.com/certbot/certbot/pull/8835#discussion_r644280409).
3. Once we've transitioned away from `patch_get_utility*`, we deprecate them and move the implementations of `get_display` and `set_display` away from zope to what you have here.

File: certbot/certbot/display/obj.py
Comment (adferrand) [reply certbot/certbot/display/obj.py]: Done.

File: certbot/certbot/tests/util.py
Comment (adferrand) [reply certbot/certbot/tests/util.py]: I was trying to imagine some workarounds involving to monkey patch `zope.component.getUtility` but sadly I do not think it will be significantly simpler that what you propose here.

Last call to see if it gives you some idea, and if not, I will implement you proposition.

Comment (bmw) [reply certbot/certbot/tests/util.py]: In an ideal world where the resources on this project were infinite, I personally like my proposition in my last post the best. It results in the design I like best in the long term without adding any cruft to our API.

Since that's not the world we live in though and refactoring all the test code around IDisplay seems like a fair bit of work and significant deviation from our initial high level goal here of adding mypy type annotations, I'm beginning to think that what you have here isn't so bad as a temporary solution. I'd ideally like to essentially implement what we discussed above in the long term, but if temporarily adding one or two functions to our public test API allows us to more easily move forward here, maybe that approach isn't so bad. I'm not sure (m)any people other than us are using this API anyway.

If you want to stick with the approach you currently have, I'd suggest:

1. Maybe renaming the functions to something like `patch_display_util*` since I think "service" is the old name.
2. Make an issue briefly explaining the test refactoring that should be done here maybe including a link to this comment thread.
3. Maybe update the warning raised out of `patch_get_utility*` to say something like "Decorator certbot.tests.util.patch_get_utility is deprecated. You should now patch certbot.display.util yourself directly or use `<new test util function name>` as a temporary workaround."
4. Update the docstring of the new functions to say something like:
```
    """Patch certbot.display.util to use a special mock IDisplay.

    The mock IDisplay works like a regular mock object, except it also
    also asserts that methods are called with valid arguments.

    The mock created by this patch mocks out Certbot internals so this can be
    used like the old patch_get_utility function. That is, the mock object will
    be called by the certbot.display.util functions and the mock returned by
    that call will be used as the IDisplay object. This was done to simplify
    the transition from zope.component and mocking certbot.display.util
    functions directly in test code should be preferred over using this
    function in the future.

    :returns: patch on the function used internally by certbot.display.util to
        get an IDisplay object
    :rtype: unittest.mock._patch

    """
```
What do you think? Either of the options described here or in my prior post work for me and I'm happy to try to help implement either if you like.

Comment (adferrand) [reply certbot/certbot/tests/util.py]: Indeed this is a realistic approach.

I made the modifications about the function names and docstring, and created the issue https://github.com/certbot/certbot/issues/8948 to keep track of the future refactoring.

File: certbot/certbot/tests/util.py
Comment (adferrand) [reply certbot/certbot/tests/util.py:220]: Regarding the discussion below, shouldn't we consider that the refactoring planned in #8948 overrides the possible changes to do on that matter in this PR?

Comment (bmw) [reply certbot/certbot/tests/util.py:220]: For the reason I described in my first post in [this thread](https://github.com/certbot/certbot/pull/8835/files#r671754133), I personally would still like to leave these functions essentially unmodified except for the warning you added.

File: certbot-apache/tests/util.py
Comment (bmw) [inline certbot-apache/tests/util.py:73]: I think using this here is a problem because we're using the internal function `certbot._internal.display.obj.set_display` across a package boundary. I think this can cause our tests to start failing before a major version bump which isn't supposed to happen.

Luckily, I think we can easily avoid this with this patch: https://gist.github.com/bbfc65284e60ae577f6929d5382f6150

File: certbot/certbot/tests/util.py
Comment (bmw) [inline certbot/certbot/tests/util.py:171]:
```suggestion
                      new_callable=_create_display_util_mock)
```

File: certbot/certbot/tests/util.py
Comment (bmw) [inline certbot/certbot/tests/util.py:201]:
```suggestion
                      new=_create_display_util_mock_with_stdout(stdout))
```

File: certbot/certbot/tests/util.py
Comment (bmw) [inline certbot/certbot/tests/util.py:307]:
```suggestion
def _create_display_util_mock():
```

File: certbot/certbot/tests/util.py
Comment (bmw) [inline certbot/certbot/tests/util.py:318]:
```suggestion
def _create_display_util_mock_with_stdout(stdout):
```

**Review by bmw — CHANGES_REQUESTED** [review]
Thanks so much for continuing to work on this!

My only suggestion other than my inline comments is to add a changelog entry or two about this. I'd recommend:

> ### Added
> 
> * New functions that Certbot plugins can use to interact with the user have
>   been added to `certbot.display.util`. We plan to deprecate using `IDisplay`
>   with `zope` in favor of these new functions in the future.
> 
> ### Changed
> 
> * The `certbot.tests.patch_get_utility*` functions have been deprecated.
>   Plugins should now patch `certbot.display.util` themselves in their tests or
>   use `certbot.tests.util.patch_display_util` as a temporary workaround.

File: certbot/certbot/_internal/cert_manager.py
Comment (adferrand) [reply certbot/certbot/_internal/cert_manager.py:56]: I don't think so. Let's revert this.

Comment (adferrand) [reply certbot/certbot/_internal/cert_manager.py:56]: Oh no I remember. This flag just does not exist, and in fact the `flag` parameter in `input` function does not even exist and is silently ignored by the rest parameter `*kwargs`. So I prefered to remove it entirely.

File: certbot-apache/tests/util.py
Comment (adferrand) [reply certbot-apache/tests/util.py]: Indeed!

File: certbot/certbot/tests/util.py
Comment (adferrand) [reply certbot/certbot/tests/util.py]: I reverted to the previous behavior.

**Comment by adferrand** [comment]
Changelog added.

File: certbot/certbot/tests/util.py
Comment (bmw) [inline certbot/certbot/tests/util.py:171]:
```suggestion
    :param str target: path to patch
```

**Review by bmw — CHANGES_REQUESTED** [review]
I noticed one small thing but other than that, I'll merge this assuming tests pass.

**Review by bmw — APPROVED** [review]

**Pull request merged**
